utils/vptr/tests/basic.cc

```diff
@@ -0,0 +1,212 @@
+      {
+        auto hostAccA = b1.get_access<sycl_acc_rw, sycl_acc_host>();
+        ASSERT_EQ(cl::sycl::codeplay::get_host_ptr_as<int>(hostAccA)[0], 1);
+      }
+      {
+        auto hostAccB = b2.get_access<sycl_acc_rw, sycl_acc_host>();
+        ASSERT_EQ(cl::sycl::codeplay::get_host_ptr_as<int>(hostAccB)[0], 2);
+      }
+    } catch (std::out_of_range e) {
+      FAIL();
+    }
+
+    ASSERT_EQ(pMap.count(), 2u);
+    SYCLfree(ptrA, pMap);
+    ASSERT_EQ(pMap.count(), 1u);
+    SYCLfree(ptrB, pMap);
+    ASSERT_EQ(pMap.count(), 0u);
+  }
+}
+
+TEST(pointer_mapper, reuse_ptr) {
+  PointerMapper pMap;
+  {
+    ASSERT_EQ(pMap.count(), 0u);
+
+    // First we insert a large buffer
+    void *initial = SYCLmalloc(100 * sizeof(int), pMap);
+    ASSERT_NE(initial, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(initial));
+    ASSERT_EQ(pMap.count(), 1u);
+
+    // Now we insert a small one, that we'll be reused
+    void *reused = SYCLmalloc(10 * sizeof(int), pMap);
+    ASSERT_NE(reused, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(reused));
+    ASSERT_EQ(pMap.count(), 2u);
+
+    // Another large buffer
+    void *end = SYCLmalloc(100 * sizeof(int), pMap);
+    ASSERT_NE(end, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(end));
+    ASSERT_EQ(pMap.count(), 3u);
+
+    // We free the intermediate one
+    SYCLfree(reused, pMap);
+    ASSERT_EQ(pMap.count(), 2u);
+
+    void *shouldBeTheSame = SYCLmalloc(10 * sizeof(int), pMap);
+    ASSERT_NE(shouldBeTheSame, nullptr);
+    ASSERT_FALSE(PointerMapper::is_nullptr(shouldBeTheSame));
+    ASSERT_EQ(pMap.count(), 3u);
+    ASSERT_EQ(shouldBeTheSame, reused);
+  }
+}
+
+TEST(pointer_mapper, multiple_alloc_free) {
+  PointerMapper pMap;
+  size_t numAllocations = (1 << 9);
+  size_t maxAllocSize = 1251;
+
+  std::random_device r;
+  std::default_random_engine e1(r());
+  std::uniform_int_distribution<int> uniform_dist(1, maxAllocSize);
+
+  {
+    int *ptrToFree = nullptr;
+
+    for (auto i = 0; i < numAllocations / 2; i++) {
+      int *current =
+          static_cast<int *>(SYCLmalloc(uniform_dist(e1) * sizeof(int), pMap));
+      // We choose a random pointer to free from the entire range
+      if (uniform_dist(e1) % 2) {
+        ptrToFree = current;
+      }
+    }
+    ASSERT_EQ(pMap.count(), (numAllocations / 2));
+    SYCLfree(ptrToFree, pMap);
+    ASSERT_EQ(pMap.count(), (numAllocations / 2) - 1);
+    int *ptrInBetween = static_cast<int *>(SYCLmalloc(50 * sizeof(int), pMap));
+    for (auto i = 0; i < numAllocations / 2; i++) {
+      int *current =
+          static_cast<int *>(SYCLmalloc(uniform_dist(e1) * sizeof(int), pMap));
+      // We choose a random pointer to free from the entire range
+      if (uniform_dist(e1) % 2) {
+        ptrToFree = current;
+      }
+    }
+    ASSERT_EQ(pMap.count(), numAllocations);
+    SYCLfree(ptrInBetween, pMap);
+    ASSERT_EQ(pMap.count(), (numAllocations - 1));
+  }
+
+  pMap.clear();
+  ASSERT_EQ(pMap.count(), 0);
```

Comment:
Since you've put it everywhere else, this 0 is missing a u - only mentioning because you were so fastidious about it everywhere else :)